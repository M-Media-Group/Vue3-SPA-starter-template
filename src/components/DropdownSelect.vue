<script setup lang="ts">
import { debounce } from "@/helpers/debounce";
import {
  filterOptions,
  orderOptionsBySelectedFirst,
} from "@/helpers/normaliseOptions";
import { useMultiselect } from "@/stories/Composables/useMultiselect";
import type { selectOption } from "@/types/listItem";
import { type PropType, computed, onMounted, ref, watch } from "vue";

const props = defineProps({
  /** The model value is an array of selected options. This element will always use an array of strings as the values, even if originally they are typed as numbers. This is to stay consistent with native HTML elements "value" behaviour. This may be changed in the future. */
  modelValue: {
    type: Array as PropType<string[]>,
    default: () => [],
  },

  /** The key to use to determine what value to return in the model */
  modelKey: {
    type: String as PropType<"id" | "render">,
    default: "id",
  },

  /** The key to use to determine the value to display */
  displayKey: {
    type: String as PropType<"id" | "render">,
    default: "render",
  },

  /** The passed options */
  options: {
    type: Array as PropType<selectOption[]>,
    default: () => [],
  },

  /** Whether the select is open or not */
  isOpen: {
    type: Boolean,
    default: false,
  },

  /** Show a select-all option */
  selectAll: {
    type: Boolean,
    default: false,
  },

  /** allow multiple selections */
  multiple: {
    type: Boolean,
    default: false,
  },

  /** The placeholder text to show */
  placeholder: {
    type: String,
    default: "Select an option",
  },

  /** Show search */
  searchable: {
    type: Boolean,
    default: false,
  },

  /** The search value. This is two-way binded so you should use a v-nodel for this to work */
  search: {
    type: String,
    default: "",
  },

  /** By default, the search will filter the options passed to it. If you are handling the search externally and don't want to filter the results locally, you can set this to false */
  searchLocally: {
    type: Boolean,
    default: true,
  },

  /** Limit visible elements. Its recommended to use this at a fairly low number and encourage the user to use the search. This limits how many HTML elements are rendered, but not how many are in the options array. */
  visibleLimit: {
    type: Number,
    default: 25,
  },

  /** Search encouragement text */
  searchEncouragement: {
    type: String,
    default: "Search to see more results",
  },

  /** Show the selected options first in the list */
  showSelectedFirst: {
    type: Boolean,
    default: false,
  },

  /** The role of the dropdown (button or undefined) */
  role: {
    type: String,
    default: undefined,
    validator: (value: string) => value === "button" || value === undefined,
  },

  /** The aria-invalid attribute */
  ariaInvalid: {
    type: Boolean,
    default: undefined,
  },

  /** The aria-busy attribute */
  ariaBusy: {
    type: Boolean,
    default: undefined,
  },

  /** Disable the entire dropdown */
  disabled: {
    type: Boolean,
    default: false,
  },

  /** If it is required to choose a value */
  required: {
    type: Boolean,
    default: false,
  },

  /** The select-all text */
  selectAllText: {
    type: String,
    default: "Select all",
  },

  /** The text to show when loading results */
  loadingText: {
    type: String,
    default: "Loading...",
  },

  /** If the search should autofocus on open */
  autofocus: {
    type: Boolean,
    default: false,
  },

  /** The placeholder for the search */
  searchPlaceholder: {
    type: String,
    default: "Search for an option",
  },
});

// Above rewritten in type declaration
const emit = defineEmits<{
  "update:modelValue": [string[]];
  "update:search": [string];
  reachedEndOfList: [];
  "update:isOpen": [boolean];
}>();

const {
  normalisedOptions,
  selecteableOptions,
  getLabel,
  isOptionSelected,
  toggleAllOptions,
  updateModelValue,
} = useMultiselect(props, emit);

const setModelValue = (event: Event) => {
  const target = event.target as HTMLInputElement;
  const value = target.value;

  updateModelValue(value, target.checked);
};

const filteredOptions = computed(() => {
  if (!normalisedOptions.value) return [];

  if (!props.searchLocally) return normalisedOptions.value;

  return filterOptions(normalisedOptions.value, props.search);
});

const orderedOptions = computed(() => {
  if (!filteredOptions.value) return [];

  if (!props.showSelectedFirst) {
    return filteredOptions.value;
  }

  return orderOptionsBySelectedFirst(
    filteredOptions.value,
    props.modelValue,
    props.modelKey
  );
});

const showSelectAll = computed(() => {
  if (!props.selectAll) return false;
  if (!props.multiple) return false;
  // If we are in search, we don't show it because we don't want to confuse the user user about what is being selected
  if (props.search) return false;
  if (!selecteableOptions.value) return false;
  return true;
});

/** Logic to setup the listening of when the user reached teh nottom of the dropdown list */
const dropdownList = ref<HTMLUListElement | null>(null);

const handleReachedEndOfList = () => {
  emit("reachedEndOfList");
};

const reachedEndOfListDebounced = debounce(
  handleReachedEndOfList,
  undefined,
  true
);

const setupDropdownList = () => {
  if (!dropdownList.value) return;

  dropdownList.value.addEventListener("scroll", () => {
    if (!dropdownList.value) return;
    if (
      dropdownList.value.scrollTop + dropdownList.value.clientHeight >=
      dropdownList.value.scrollHeight
    ) {
      reachedEndOfListDebounced();
    }
  });
};

const searchInput = ref<HTMLInputElement | null>(null);

/** A dunction to handle the opening of the results. If the search is present, we autofocus the search input. Note, there is a `ToggleEvent` type that should work but doesnt pass the TS linting with `TS2304: Cannot find name 'ToggleEvent'.`. */
const openResults = async (value: { newState: string }) => {
  if (!value) return;

  // If the value is not the same as the current value, emit the event
  const valueCompare = value.newState === "open";

  if (valueCompare !== props.isOpen) {
    emit("update:isOpen", valueCompare);
  }
};

onMounted(() => {
  setupDropdownList();
});

const getSummaryText = () => {
  if (props.modelValue.length > 0) {
    return props.modelValue
      .map((value) => {
        const option = normalisedOptions.value.find((option) =>
          typeof option === "string"
            ? option === value
            : option[props.modelKey] === value
        );
        return option
          ? getLabel(option)
          : value.trim() !== ""
          ? value
          : props.placeholder;
      })
      .join(", ");
  }

  return props.placeholder;
};

const canBeFocusedOn = computed(
  () =>
    (props.isOpen &&
      searchInput.value &&
      props.searchable &&
      props.autofocus) ??
    false
);

watch(
  () => props.isOpen,
  () => {
    if (!canBeFocusedOn.value) return;
    searchInput.value?.focus();
    // Scroll the dropdown to the top - this is needed because the CSS height animation seems to leave us a little bit scrolled down
    if (dropdownList.value) {
      dropdownList.value.scrollTop = 0;
    }
  }
);

// Watch the search for changes and scroll up if needed
watch(
  () => props.search,
  () => {
    if (dropdownList.value) {
      dropdownList.value.scrollTop = 0;
    }
  }
);

// Define a focus function that can be called when we use the ref on this component
const focus = () => {
  // If the dropdown is not open, open it
  if (!props.isOpen) {
    emit("update:isOpen", true);
  }
};

// Expose the focus function to the parent component
defineExpose({ focus });
</script>
<template>
  <details class="dropdown" :open="props.isOpen" @toggle="openResults">
    <summary
      :role="props.role"
      :aria-invalid="props.ariaInvalid"
      :aria-busy="props.ariaBusy"
      :data-has-value="props.modelValue.length > 0"
    >
      {{ props.modelValue.length > 0 ? getSummaryText() : props.placeholder }}
    </summary>
    <ul ref="dropdownList">
      <li v-if="props.searchable">
        <input
          type="search"
          :value="props.search"
          @input="
            $emit('update:search', ($event.target as HTMLInputElement).value)
          "
          :aria-busy="props.ariaBusy"
          :placeholder="props.searchPlaceholder"
          :aria-label="props.searchPlaceholder"
          :autofocus="canBeFocusedOn"
          ref="searchInput"
        />
      </li>
      <li v-if="showSelectAll">
        <label>
          <input
            type="checkbox"
            :checked="props.modelValue.length === selecteableOptions?.length"
            @click="toggleAllOptions"
            :disabled="props.disabled"
            value="all"
            tabindex="0"
          />
          {{ props.selectAllText }}
        </label>
      </li>

      <li
        v-for="option in orderedOptions?.slice(0, props.visibleLimit)"
        :key="option[modelKey]"
      >
        <slot
          name="optionSlot"
          :option="option"
          :checked="isOptionSelected(option)"
          :updateModelValue="setModelValue"
          :modelValue="props.modelValue"
        >
          <label>
            <input
              type="checkbox"
              :disabled="option.disabled || props.disabled"
              :value="option[modelKey]"
              :checked="isOptionSelected(option)"
              @click="setModelValue"
              tabindex="0"
            />
            {{ getLabel(option) }}
          </label>
        </slot>
      </li>

      <li v-if="ariaBusy" aria-busy="true">{{ props.loadingText }}</li>
      <!-- if the results are limited because of the visible limit, but there are more filteredResults, show a message -->
      <li v-if="filteredOptions.length > props.visibleLimit">
        {{ props.searchEncouragement }}
      </li>
    </ul>
  </details>
</template>
<style scoped>
ul {
  max-height: 400px;
  overflow-y: auto;
  list-style: none;
}
</style>
