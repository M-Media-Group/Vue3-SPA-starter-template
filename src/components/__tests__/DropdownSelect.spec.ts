import { describe, expect, it } from "vitest";

import { RouterLinkStub, mount } from "@vue/test-utils";
import DropdownSelect from "../DropdownSelect.vue";

describe("Dropdown Select", () => {
  it("renders correctly", async () => {
    const start = performance.now();

    const wrapper = mount(DropdownSelect, {
      props: {
        options: ["One", "Two", "Three"],
        "onUpdate:modelValue": (value) => {
          wrapper.setProps({ modelValue: value });
        },
        isOpen: false,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const end = performance.now();

    expect(end - start).toBeLessThan(50); // should be set to 25 but it makes test flakey

    // There should be a select element
    const select = wrapper.find("summary");
    expect(select.exists()).toBe(true);

    // There should be 3 options
    const options = wrapper.findAll("li");
    expect(options.length).toBe(3);

    // The options should have the correct text
    expect(options[0].text()).toBe("One");
    expect(options[1].text()).toBe("Two");
    expect(options[2].text()).toBe("Three");

    // Expect the snapshot to match
    expect(wrapper.html()).toMatchSnapshot();

    // There should not be a search input
    const input = wrapper.find("input[type='search']");
    expect(input.exists()).toBe(false);
    expect(wrapper.emitted("update:search")).toBeFalsy();

    // There should not be a select all checkbox
    const selectAll = wrapper.find("input[type='checkbox'][value='all']");
    expect(selectAll.exists()).toBe(false);

    // By default the results should not be open
    const details = wrapper.find("details");
    expect(details.attributes("open")).toBe(undefined);
    expect(wrapper.emitted("update:isOpen")).toBeFalsy();

    // Clicking on the summary should open the dropdown
    await select.trigger("click");
    expect(details.attributes("open")).toBe("");

    // Ensure the prop value is updated
    // expect(wrapper.emitted("update:isOpen")).toBeTruthy();
    // expect(wrapper.props("isOpen")).toBe(true);

    // The default placeholder should be "Select an option"
    expect(select.text()).toBe("Select an option");

    // There should be 3 labels
    const labels = wrapper.findAll("label");
    expect(labels.length).toBe(3);

    // Click on the second option
    await labels[1].trigger("click");

    expect(wrapper.emitted("update:modelValue")?.[0]).toEqual([["Two"]]);

    // There should be no placeholder
    expect(select.text()).toBe("Two");

    // Clicking the same option again should uncheck it and set the placeholder back
    await labels[1].trigger("click");
    expect(wrapper.emitted("update:modelValue")?.[1]).toEqual([[]]);
    expect(select.text()).toBe("Select an option");

    // When we add more options, the dropdown should still render
    await wrapper.setProps({ options: ["One", "Two", "Three", "Four"] });

    // There should be 4 options
    const newOptions = wrapper.findAll("li");
    expect(newOptions.length).toBe(4);

    // When the value is set to something that is not in the options, the placeholder should show the value
    await wrapper.setProps({ modelValue: ["Five"] });
    expect(select.text()).toBe("Five");

    // If no options are passed and the value is an empty string, the placeholder should be "Select an option"
    await wrapper.setProps({ options: [], modelValue: [""] });
    expect(select.text()).toBe("Select an option");
  });

  it("emits the correct value when an option is selected", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        open: true,
        options: ["One", "Two", "Three"],
        modelValue: ["One"],
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // Click on the summary to open the dropdown
    await wrapper.find("details").trigger("click");

    const details = wrapper.find("details");

    // Assert there is the open attribute
    expect(details.attributes("open")).toBe("");

    // Assure 3 options are visible
    const options = wrapper.findAll("input");
    expect(options.length).toBe(3);

    // There should be 3 labels
    const labels = wrapper.findAll("label");
    expect(labels.length).toBe(3);

    // Click on the second option
    await labels[1].trigger("click");

    // Expect the correct value to be emitted
    expect(wrapper.emitted("update:modelValue")).toBeTruthy();
  });

  it("shows the search input when the searchable prop is true", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        searchable: true,
        searchPlaceholder: "Search for something",
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be a search input
    const input = wrapper.find("input[type='search']");
    expect(input.exists()).toBe(true);

    // By default the input should not have any focus
    expect(input.element).not.toBe(document.activeElement);

    // Opening the dropdown should focus the input
    await wrapper.find("details").trigger("click");
    // expect(input.element).toBe(document.activeElement);

    // Typing in the seach should emit the search event
    input.setValue("Some value");
    expect(wrapper.emitted("update:search")).toBeTruthy();

    // The search should have a placeholder of "Search for something"
    expect(input.attributes("placeholder")).toBe("Search for something");
  });

  it("correctly filters results when the search prop is used", () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        searchable: true,
        options: ["One", "Two", "Three"],
        search: "One",
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });
    // There should be 1 option visible
    const options = wrapper.findAll("label");
    expect(options[0].text()).toBe("One");
    expect(options.length).toBe(1);
  });

  it("does not filter the results when searchLocally is false", () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        searchable: true,
        searchLocally: false,
        options: ["One", "Two", "Three"],
        search: "One",
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });
    // There should be 3 options visible
    const options = wrapper.findAll("label");
    expect(options.length).toBe(3);
  });

  it("correctly filters results when the search prop is used by typing", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        searchable: true,
        options: ["One", "Two", "Three"],
        search: "",
        "onUpdate:search": (value) => {
          wrapper.setProps({ search: value });
        },
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const input = wrapper.find("input[type='search']");
    await input.setValue("Two");

    expect(wrapper.emitted("update:search")?.[0]).toEqual(["Two"]);
    expect(wrapper.props("search")).toBe("Two");

    // There should be 1 option visible
    const options = wrapper.findAll("label");
    // There should only be one option visible
    expect(options.length).toBe(1);

    expect(options[0].text()).toBe("Two");
  });

  it("correctly displays options when they are passed as objects", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        open: true,
        multiple: true,
        options: [
          {
            id: 1,
            render: "Hello world",
          },
          {
            id: "2",
            render: "This computed callback",
          },
          {
            id: 3,
            render: "Something else",
          },
        ],
        modelValue: [],
        "onUpdate:modelValue": (value) => {
          wrapper.setProps({ modelValue: value });
        },
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // Click on the summary to open the dropdown
    await wrapper.find("details").trigger("click");

    const details = wrapper.find("details");

    // Assert there is the open attribute
    expect(details.attributes("open")).toBe("");

    // Assure 3 options are visible
    const options = wrapper.findAll("input");
    expect(options.length).toBe(3);

    // There should be 3 labels
    const labels = wrapper.findAll("label");
    expect(labels.length).toBe(3);

    // Click on the second option
    await labels[1].trigger("click");

    // Expect the correct value to be emitted
    expect(wrapper.emitted("update:modelValue")?.[0]).toEqual([["2"]]);

    // The summary label should be the first option (its render property)
    expect(wrapper.find("summary").text()).toBe("This computed callback");

    // Selecting another option should add it to the modelValue
    await labels[0].trigger("click");

    // Expect the correct value to be emitted
    expect(wrapper.emitted("update:modelValue")?.[1]).toEqual([["2", "1"]]);

    // Click summary label should have the correct text
    expect(wrapper.find("summary").text()).toBe(
      "This computed callback, Hello world"
    );

    // Unchecking the second option should remove it from the modelValue
    await labels[1].trigger("click");

    // Expect the correct value to be emitted
    expect(wrapper.emitted("update:modelValue")?.[2]).toEqual([["1"]]);

    // Click summary label should have the correct text
    expect(wrapper.find("summary").text()).toBe("Hello world");

    // Unchecking the first option should remove it from the modelValue
    await labels[0].trigger("click");

    // Expect the correct value to be emitted
    expect(wrapper.emitted("update:modelValue")?.[3]).toEqual([[]]);

    // Click summary label should have the correct text
    expect(wrapper.find("summary").text()).toBe("Select an option");
  });

  it("correctly displays options when they are passed as slots into the `optionSlot` slot", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        open: true,
        options: ["One", "Two", "Three"],
        modelValue: ["One"],
      },
      slots: {
        // The slot is called `option` and it receives the option as a prop
        optionSlot: `
          <template #option="{ option, updateModelValue, modelValue }">
            <label>
              <input type="checkbox" :checked="modelValue.includes(option)" @click="updateModelValue" />
              Hello - {{ option.render }}
            </label>
          </template>
        `,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // Click on the summary to open the dropdown
    await wrapper.find("details").trigger("click");

    const details = wrapper.find("details");

    // Assert there is the open attribute
    expect(details.attributes("open")).toBe("");

    // Assure 3 options are visible
    const options = wrapper.findAll("input");
    expect(options.length).toBe(3);

    // There should be 3 labels
    const labels = wrapper.findAll("label");
    expect(labels.length).toBe(3);

    // Each option should have "Hello - " prepended to it
    expect(labels[0].text()).toBe("Hello - One");

    // Click on the second option
    await labels[0].trigger("click");

    // Expect the correct value to be emitted
    expect(wrapper.emitted("update:modelValue")).toBeTruthy();
  });

  it("emits an event when the user reached the bottom of the dropdown options list", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        open: true,
        // Generate 50 options
        options: Array.from({ length: 50 }, (_, i) => `Option ${i}`),
        modelValue: ["One"],
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // Initially, the event should not be emitted
    expect(wrapper.emitted("reachedEndOfList")).toBeFalsy();

    // Click on the summary to open the dropdown
    await wrapper.find("details").trigger("click");

    // Scroll to the bottom of the dropdown
    await wrapper.find("ul").trigger("scroll");

    // Expect the event to be emitted
    expect(wrapper.emitted("reachedEndOfList")?.length).toBe(1);
  });

  it("allows select all and deselect all when the `selectAll` prop is true", async () => {
    const wrapper = mount(DropdownSelect, {
      props: {
        open: true,
        options: ["One", "Two", "Three"],
        selectAll: true,
        multiple: true,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // Click on the summary to open the dropdown
    await wrapper.find("details").trigger("click");

    // Assert there's a Select all checkbox. It will have a value of all
    const selectAll = wrapper.find("input[type='checkbox'][value='all']");
    expect(selectAll.exists()).toBe(true);

    // Click on the select all checkbox which will be the first one in the list
    await wrapper.find("input[type='checkbox']").trigger("click");

    // Expect all options to be selected
    expect(wrapper.emitted("update:modelValue")?.[0]).toEqual([
      ["One", "Two", "Three"],
    ]);

    // Click on the select all checkbox again
    await wrapper.find("input[type='checkbox']").trigger("click");

    // Expect all options to be deselected
    expect(wrapper.emitted("update:modelValue")?.[1]).toEqual([[]]);
  });

  it("should render a large number of options, 1000000, in less than 400ms", () => {
    const options = Array.from({ length: 1000000 }, (_, i) => `Option ${i}`);
    const start = performance.now();
    const wrapper = mount(DropdownSelect, {
      props: {
        options,
        visibleLimit: 25,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });
    const end = performance.now();
    expect(end - start).toBeLessThan(2000); // should be set to 500 but it makes test flakey

    // There should be the limited number of options visible (25)
    const visibleOptions = wrapper.findAll("label");
    expect(visibleOptions.length).toBe(25);
  });
});
