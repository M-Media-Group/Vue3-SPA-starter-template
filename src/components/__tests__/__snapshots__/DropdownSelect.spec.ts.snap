// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown Select > renders correctly 1`] = `
"<details data-v-10d9b766="" class="dropdown">
  <summary data-v-10d9b766="" data-has-value="false">Select an option</summary>
  <ul data-v-10d9b766="">
    <!--v-if-->
    <!--v-if-->
    <li data-v-10d9b766=""><label data-v-10d9b766=""><input data-v-10d9b766="" type="checkbox" tabindex="0" value="One"> One</label></li>
    <li data-v-10d9b766=""><label data-v-10d9b766=""><input data-v-10d9b766="" type="checkbox" tabindex="0" value="Two"> Two</label></li>
    <li data-v-10d9b766=""><label data-v-10d9b766=""><input data-v-10d9b766="" type="checkbox" tabindex="0" value="Three"> Three</label></li>
    <!--v-if-->
    <!-- if the results are limited because of the visible limit, but there are more filteredResults, show a message -->
    <!--v-if-->
  </ul>
</details>"
`;
